camel-idea-plugin/src/main/java/org/apache/camel/idea/util/CamelService.java

```diff
@@ -100,18 +116,150 @@ public void scanForCamelDependencies(@NotNull Module module) {
+                    }
                 }
             }
         }
     }
 
-    @Override
-    public void dispose() {
-        processedLibraries.clear();
+    /**
+     * Scan for Custom Camel Libraries and update the cache and camel catalog with custom components discovered
+     */
+    public void scanForCustomCamelDependencies(@NotNull Project project, @NotNull Module module) {
+        CamelCatalog camelCatalog = ServiceManager.getService(project, CamelCatalogService.class).get();
+
+        for (OrderEntry entry : ModuleRootManager.getInstance(module).getOrderEntries()) {
+            if (entry instanceof LibraryOrderEntry) {
+                LibraryOrderEntry libraryOrderEntry = (LibraryOrderEntry) entry;
+
+                String name = libraryOrderEntry.getPresentableName().toLowerCase();
+                if (libraryOrderEntry.getScope().isForProductionCompile() || libraryOrderEntry.getScope().isForProductionRuntime()) {
+                    final Library library = libraryOrderEntry.getLibrary();
+                    if (library == null) {
+                        continue;
+                    }
+                    String[] split = name.split(":");
+                    String groupId = split[1].trim();
+                    String artifactId = split[2].trim();
+
+                    // skip reserved from Apache Camel itself
+                    if ("org.apache.camel".equals(groupId)) {
+                        continue;
+                    }
+
+                    try (URLClassLoader classLoader = newURLClassLoaderForLibrary(library)) {
+                        if (classLoader != null) {
+                            // is there any custom Camel components in this library?
+                            Properties properties = loadComponentProperties(classLoader);
+                            if (properties != null) {
+                                String components = (String) properties.get("components");
+                                if (components != null) {
+                                    String[] part = components.split("\\s");
+                                    for (String scheme : part) {
+                                        if (!camelCatalog.findComponentNames().contains(scheme)) {
+                                            // find the class name
+                                            String javaType = extractComponentJavaType(classLoader, scheme);
+                                            if (javaType != null) {
+                                                String json = loadComponentJSonSchema(classLoader, scheme);
+                                                if (json != null) {
+                                                    camelCatalog.addComponent(scheme, javaType, json);
+                                                    // okay a new Camel component was added
+                                                    if (!containsLibrary(artifactId)) {
```

Comment:
Wouldn't it make sense to also do this check before all the above code and then continue if it already exists in the library ? It possible you will scan the same library because modules share libraries